src/Datadog.Trace/AppSec/AspNetCore/BlockingMiddleware.cs

```diff
@@ -0,0 +1,91 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#if !NETFRAMEWORK
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Logging;
+using Microsoft.AspNetCore.Http;
+
+namespace Datadog.Trace.AppSec.AspNetCore
+{
+    internal static class BlockingMiddleware
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(BlockingMiddleware));
+
+        public static void ModifyASpplicationBuilder(object applicationBuilderInstance)
+        {
+            if (applicationBuilderInstance.TryDuckCast<ApplicationBuilderDuck>(out var applicationBuilder))
+            {
+                InsertMiddlewares(applicationBuilder.Components);
+            }
+            else
+            {
+                Log.Error($"Couldn't create duck type from {applicationBuilderInstance?.GetType()?.FullName ?? "(null)"}");
+            }
+        }
+
+        private static void InsertMiddlewares(List<Func<RequestDelegate, RequestDelegate>> components)
+        {
+            static async Task Middleware(HttpContext context, Func<Task> next)
+            {
+                if (context.Items.ContainsKey(SecurityConstants.KillKey) && context.Items[SecurityConstants.KillKey] is bool killKey && killKey)
+                {
+                    await BlockRequest(context);
+                }
+                else
+                {
+                    try
+                    {
+                        context.Items[SecurityConstants.InHttpPipeKey] = true;
+                        await next.Invoke();
+                    }
+                    catch (BlockActionException)
+                    {
+                        await BlockRequest(context);
+                    }
+                }
+            }
+
+            RequestDelegate MiddlewareWrapper(RequestDelegate next) => context => Middleware(context, () => next(context));
+
+            components.Insert(0, MiddlewareWrapper);
+            if (components.Count > 2)
+            {
+                // insert 2nd to last, making a guess that the last one will be the user action
+                components.Insert(components.Count - 2, MiddlewareWrapper);
+            }
+        }
+
+        private static async Task BlockRequest(HttpContext context)
+        {
+            context.Response.StatusCode = 403;
+            context.Response.ContentType = "text/html";
+            await context.Response.WriteAsync(SecurityConstants.AttackBlockedHtml);
+            var callbackExists = context.Items.TryGetValue("Security", out var result);
+            if (callbackExists && result is Guid callbackId)
+            {
+                Security.Instance.Execute(callbackId);
+            }
+        }
```

Comment:
I can't see any easy way to check if the headers have been sent using the .NET Standard ASP.NET Core libraries we have available, but maybe I'm missing something.

I think it something we want to do, but I think handling can wait because:
- it's unlikely clients will activate blocking for a while because want to see it running in report mode till they're more confident
- if headers have been sent, we probably want throw an exception to stop further data being sent.

I've logged https://datadoghq.atlassian.net/browse/APPSEC-985 to track this.

Reply:
I believe you can use `context.Response.HasStarted` to detect that - I agree it's complex what to do, but if we leave this as-is, users will see `InvalidOperationException` which could be confusing. Better to detect it and throw a security specific exception I think 🙂

Reply:
Okay, done in a1feec09